core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -381,21 +385,29 @@ class KafkaApis(val requestChannel: RequestChannel,
       return
     }
 
-    val (existingAndAuthorizedForDescribeTopics, nonExistingOrUnauthorizedForDescribeTopics) =
-      produceRequest.partitionRecordsOrFail.asScala.partition { case (tp, _) =>
-        authorize(request.session, Describe, new Resource(Topic, tp.topic)) && metadataCache.contains(tp.topic)
-      }
+    var unauthorizedTopics = Set[TopicPartition]()
+    var nonExistingAndAuthorizedForDescribeTopics = Set[TopicPartition]()
+    var authorizedRequestInfo = mutable.Map[TopicPartition, MemoryRecords]()
 
-    val (authorizedRequestInfo, unauthorizedForWriteRequestInfo) = existingAndAuthorizedForDescribeTopics.partition {
-      case (tp, _) => authorize(request.session, Write, new Resource(Topic, tp.topic))
+    for ((topicPartition, memoryRecords) <- produceRequest.partitionRecordsOrFail.asScala) {
+      if (authorize(request.session, Describe, new Resource(Topic, topicPartition.topic)))
```

Comment:
Thinking about ways to condense these checks. At least maybe we could flatten it?
```
if (!authorize(Describe))
  TOPIC_AUTH
else if (!metadata.contains)
  UNKNOWN
else if (!authorize(Write))
  TOPIC_AUTH
else
  GOOD
```

Reply:
Briefly discussed this with @ijuma offline, but in some of these cases, I'm not sure the `Describe` check is even needed.  Is it useful to be able to distinguish the UNKNOWN_TOPIC case if the user doesn't have `Write` access anyway? Maybe we could simplify this a little:
```
if (!authorize(Write))
  TOPIC_AUTH
else if (!metadata.contains)
  UNKNOWN
else
  GOOD
```
Thoughts?

Reply:
Agree.  since (Read, Write, Delete, Alter) operations implies describe operation, It makes sense to do single authorize check at the beginning and then do topic existence check.

Updated the PR. Please check.